vms/proposervm/state/chain_state.go

```diff
@@ -22,15 +26,67 @@ type ChainState interface {
 	SetLastAccepted(blkID ids.ID) error
 	DeleteLastAccepted() error
 	GetLastAccepted() (ids.ID, error)
+
+	PutVerifiedBlock(blkID ids.ID) error
+	DeleteVerifiedBlock(blkID ids.ID) error
+	SetPreference(preferredID ids.ID) error
+	Preferred() ids.ID
 }
 
 type chainState struct {
 	lastAccepted ids.ID
+	preferred    ids.ID
 	db           database.Database
 }
 
-func NewChainState(db database.Database) ChainState {
-	return &chainState{db: db}
+func NewChainState(db database.Database, blkState BlockState) (ChainState, error) {
+	s := &chainState{db: db}
+	return s, s.initPreferredChain(blkState)
+}
+
+func (s *chainState) initPreferredChain(blkState BlockState) error {
+	preferredID, err := s.db.Get(preferredKey)
+	switch {
+	case err != nil && err != database.ErrNotFound:
+		return err
+	case err == database.ErrNotFound:
+		return nil
+	}
+
+	// Add the preferred chain up to the last accepted block to be skipped.
+	preferredBlkIDs := make(map[string]struct{})
+	preferredBlk, status, err := blkState.GetBlock(ids.ID(preferredID))
+	if err != nil {
+		return err
+	}
+	for status != choices.Accepted {
+		preferredID := preferredBlk.ID()
+		preferredBlkIDs[string(preferredID[:])] = struct{}{}
+
+		preferredBlk, status, err = blkState.GetBlock(preferredBlk.ParentID())
+		if err != nil {
+			return err
+		}
+	}
+
+	// Delete any verified blocks that are not in the preferred chain.
```

Comment:
Do we have to do a sweep here? I thought it might make sense for us to just rely on deleting these verified blocks upon `Accept` and `Reject` since we know that one will eventually be called when issuing to consensus + we call verify before consensus to insert.

Reply:
We only maintain that invariant when the node is online. When the node turns off with blocks that have been verified/inserted into consensus, we do not maintain the invariant that Accept/Reject will be called after restart. That's the reason we add the extra layer of tracking here to ensure they're still cleaned up properly.

The consensus engine doesn't currently persist anything about its preferences, confidence counters, or what blocks have been verified/accepted/rejected. It relies entirely on the VM for this.